src/services/fileServices/MdPageServices/PageService.ts

```diff
@@ -0,0 +1,379 @@
+// this should return a sum type of the possible errors.
+// This is (at a glance), either a failure to read from github
+// or the parsing fails (we have a sentinel value of an empty string)
+const withError = (promise: Promise<PageInfo>) =>
+  ResultAsync.fromPromise(promise, () => new BaseIsomerError()).andThen(
+    (data) =>
+      // NOTE: This is a fail-safe check as `yaml.parse`
+      // doesn't guarantee existence.
+      data?.content?.frontMatter?.permalink
+        ? ok(data)
+        : err(new BaseIsomerError())
+  )
+
+const extractStagingPermalink = (stagingLink: string) => ({
+  content,
+}: PageInfo): StagingPermalink =>
+  Brand.fromString(`${stagingLink}${content.frontMatter.permalink}`)
+
+interface PageServiceProps {
+  contactUsService: ContactUsPageService
+  collectionPageService: CollectionPageService
+  subCollectionPageService: SubcollectionPageService
+  homepageService: HomepagePageService
+  resourcePageService: ResourcePageService
+  unlinkedPageService: UnlinkedPageService
+  resourceRoomDirectoryService: ResourceRoomDirectoryService
+}
+
+// eslint-disable-next-line import/prefer-default-export
+export class PageService {
+  private contactUsService: ContactUsPageService
+
+  private collectionPageService: CollectionPageService
+
+  private subCollectionPageService: SubcollectionPageService
+
+  private homepageService: HomepagePageService
+
+  private resourcePageService: ResourcePageService
+
+  private unlinkedPageService: UnlinkedPageService
+
+  private resourceRoomDirectoryService: ResourceRoomDirectoryService
+
+  constructor({
+    contactUsService,
+    collectionPageService,
+    subCollectionPageService,
+    homepageService,
+    resourcePageService,
+    unlinkedPageService,
+    resourceRoomDirectoryService,
+  }: PageServiceProps) {
+    this.contactUsService = contactUsService
+    this.collectionPageService = collectionPageService
+    this.subCollectionPageService = subCollectionPageService
+    this.homepageService = homepageService
+    this.resourcePageService = resourcePageService
+    this.unlinkedPageService = unlinkedPageService
+    this.resourceRoomDirectoryService = resourceRoomDirectoryService
+  }
+
+  /**
+   * This method assumes that only 1 call to `_config.yml` is needed
+   * to determine a page's name.
+   *
+   * In order for this assumption to hold true, we maintain the invariant that
+   * the method call **must always** act upon some existing fully qualified file path from
+   * the frontend.
+   *
+   * This is done to avoid expensive calls to fetch a page's raw blob data and
+   * parsing it thereafter which, for big files, might be expensive.
+   * @param pageName A **valid and fully qualified** file path
+   * @param sessionData session credentials of the user
+   */
+  parsePageName = (
+    pageName: string,
+    sessionData: UserSessionData
+  ): ResultAsync<PageName, NotFoundError | MissingResourceRoomError> =>
```

Comment:
could we just use `NotFoundError` rather than `NotFoundError | MissingResourceRoomError` as `MissingResourceRoomError` is a type of `NotFoundError`?

Reply:
no, that's actually discouraged. we want to return the **most specific** type that we can. Note that in this case, for eg, by returning `NotFoundError | MissingResourceRoomError`, the caller can determine whether they care about the _type_ of error. 

If we were to return a blanket `NotFoundError`, the caller has to perform validation checks to see if the error is indeed something that they care about. If the caller does **not** care about it, they can always just use the properties etc and we don't lose anything if the subsequent method call requires a `NotFoundError` (cos the interface is the same).

Reply:
hmm how would the caller determine if it's indeed a missing resource room? the exact nature of the error would be hidden behind a general interface of `NotFoundError`, no? Esp if, for eg (as per this endpoint), the call can return many various kinds of not found error

Reply:
uhm ok wait to be more specific, my question would be wouldnt the cast 
`NotFoundError | MissingResourceRoomError` be semantically equivalent to `NotFoundError` as `MissingResourceRoomError` _extends_ from `NotFoundError`?

Reply:
Actually because ts uses structural typing, all the errors without differing properties will be identical. This means, for example, that we can even assign a totally unrelated error (`AuthError`) to a method/variable that requires `MissingSiteError`/`NotFoundError`.

what we want to think about here is whether the error itself is useful - we can return a `NotFoundError` but we could be more specific and return a `MissingResourceRoomError`. This allows the caller to tell what errors are returned from thsi function.

as maintainers, we can also know the specific `NotFound` errors that can be returned (eg, if the file wasn't actually found on github, if the contact-us page is missing etc), which helps with ease of maintainability

Reply:
arh ok this line of thought makes sense, thanks for the explanation!